src/base/cub_ddl_log.c

```diff
@@ -0,0 +1,1054 @@
+
+  if (retval < 0)
+    {
+      assert (false);
+      dest_path[0] = '\0';
+      return retval;
+    }
+#if defined(WINDOWS)
+  unix_style_path (dest_path);
+#endif
+
+  tpath = strdup (dest_path);
+  if (tpath == NULL)
+    {
+      return -1;
+    }
+  cub_create_dir_log (tpath);
+  FREE (tpath);
+
+  return retval;
+}
+
+static int
+cub_make_ddl_log_filename (char *filename_buf, size_t buf_size, const char *app_name)
+{
+  const char *env_root = NULL;
+  int retval = 0;
+
+  assert (filename_buf != NULL);
+
+  env_root = envvar_root ();
+
+  if (strcmp (app_name, "cas") == 0)
+    {
+      retval =
+	snprintf (filename_buf, buf_size, "%s/%s/%s_%d.ddl.log", env_root, DDL_LOG_PATH, ddl_audit_handle->br_name,
+		  (ddl_audit_handle->br_index + 1));
+    }
+  else if (strcmp (app_name, "csql") == 0 || strcmp (app_name, "loaddb") == 0)
+    {
+      retval =
+	snprintf (filename_buf, buf_size, "%s/%s/%s_%s_ddl.log", env_root, DDL_LOG_PATH, app_name,
+		  ddl_audit_handle->db_name);
+    }
+  else
+    {
+      filename_buf[0] = '\0';
+      return -1;
+    }
+
+  if (retval < 0)
+    {
+      assert (false);
+      filename_buf[0] = '\0';
+    }
+#if defined(WINDOWS)
+  unix_style_path (filename_buf);
+#endif
+  return retval;
+}
+
+static FILE *
+cub_ddl_log_open (char *app_name)
+{
+  FILE *fp = NULL;
+  char *tpath = NULL;
+  int len;
+
+  if (app_name != NULL)
+    {
+      len = cub_make_ddl_log_filename (ddl_audit_handle->log_filepath, PATH_MAX, app_name);
+
+      if (ddl_audit_handle->log_filepath[0] == '\0' || len < 0)
+	{
+	  goto file_error;
+	}
+
+      tpath = strdup (ddl_audit_handle->log_filepath);
+      if (tpath == NULL)
+	{
+	  goto file_error;
+	}
+      cub_create_dir_log (tpath);
+      FREE_MEM (tpath);
+    }
+  else
+    {
+      goto file_error;
+    }
+
+  /* note: in "a+" mode, output is always appended */
+  fp = cub_fopen_and_lock (ddl_audit_handle->log_filepath, "a+");
+
+  if (fp != NULL)
+    {
+      fseek (fp, 0, SEEK_END);
+    }
+  return fp;
+
+file_error:
+  return NULL;
+}
+
+void
+cub_ddl_log_write_end ()
+{
+  if (ddl_audit_handle == NULL)
+    {
+      return;
+    }
+
+  if (prm_get_bool_value (PRM_ID_DDL_AUDIT_LOG) == FALSE)
+    {
+      goto ddl_log_free;
+    }
+
+  if (strcmp (ddl_audit_handle->app_name, "loaddb") == 0)
+    {
+      if (ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_SCHEMA)
+	{
+	  goto ddl_log_free;
+	}
+    }
+  else
+    {
+      if (cub_is_ddl_type (ddl_audit_handle->stmt_type) == FALSE)
+	{
+	  goto ddl_log_free;
+	}
+    }
+
+  if (ddl_audit_handle->log_type & DDL_LOG_ELAPSED_TIME)
+    {
+      cub_ddl_log_write ();
+    }
+
+ddl_log_free:
+  cub_ddl_log_free ();
+}
+
+void
+cub_ddl_log_write ()
+{
+  FILE *fp = NULL;
+  char buf[DDL_LOG_BUFFER_SIZE] = { 0 };
+  char dest_path[PATH_MAX] = { 0 };
+  int len = 0;
+  int ret = 0;
+
+  if (ddl_audit_handle == NULL)
+    {
+      return;
+    }
+
+  if (prm_get_bool_value (PRM_ID_DDL_AUDIT_LOG) == FALSE)
```

Comment:
Minor suggestion: (2 if-statements in one)
`if (ddl_audit_handle == NULL || prm_get_bool_value (PRM_ID_DDL_AUDIT_LOG) == FALSE)`